algorithms/color_refinement.py

```diff
@@ -0,0 +1,69 @@
+from supporting_components.graph import Graph, Vertex
+from typing import List
+
+
+def color_refinement(G: "Graph"):
+    """
+    Assuming there is no initial coloring
+    :param G: Graph to be colored
+    :return: Colored graph
+    """
+
+    colors, max_colornum = __initialize_colors(G)
+
+    has_changed = True
+    while has_changed:
+        for color_num, vertices in colors.copy().items():
+            if has_changed:
+                max_colornum += 1
+                has_changed = False
+
+            if len(vertices) == 1:
+                continue
+
+            has_changed = __colorgroup_refinement(colors, vertices, max_colornum)
+    return G
+
+
+def __initialize_colors(G: "Graph"):
+    colors = {}
+    max_colornum = 0
+    for v in G.vertices:
+        v.colornum = v.degree
+        if v.degree > max_colornum:
+            max_colornum = v.degree
+        colors.setdefault(v.degree, []).append(v)
+    return colors, max_colornum
+
+
+def __colorgroup_refinement(colors, vertices: List["Vertex"], max_colornum):
+    """
+
```

Comment:
Hier ook uitleg van de functie, ook dat de groep in maximaal twee nieuwe groepen wordt verdeeld.